packages/backend-common/src/service/lib/ServiceBuilderImpl.ts

```diff
@@ -121,20 +145,91 @@ export class ServiceBuilderImpl implements ServiceBuilder {
         reject(e);
       });
 
-      const server = stoppable(
-        app.listen(port, host, () => {
-          logger.info(`Listening on ${host}:${port}`);
-        }),
-        0,
-      );
+      let server: http.Server;
+
+      if (httpsSettings) {
+        logger.info('Initializing https server');
+
+        const credentials: { key: string; cert: string } = {
+          key: '',
+          cert: '',
+        };
+        const signingOptions: any = httpsSettings?.certificate;
+
+        if (signingOptions?.algorithm !== undefined) {
+          logger.info('Generating self-signed certificate with attributes');
+
+          const certificateAttributes: Array<any> = Object.entries(
+            signingOptions.attributes,
+          ).map(([name, value]) => ({ name, value }));
+
+          // TODO: Create a type def for selfsigned.
+          const signatures = require('selfsigned').generate(
+            certificateAttributes,
+            {
+              algorithm: signingOptions?.algorithm,
+              keySize: signingOptions?.size || 2048,
+              days: signingOptions?.days || 30,
+            },
+          );
+
+          logger.info(
+            'Bootstrapping key and cert from self-signed certificate',
+          );
+
+          credentials.key = signatures.private;
+          credentials.cert = signatures.cert;
+        } else {
+          if (fs.existsSync(signingOptions?.key)) {
+            if (fs.lstatSync(signingOptions?.key).isFile()) {
+              logger.info('Bootstrapping key from file');
+
+              credentials.key = fs.readFileSync(signingOptions?.key).toString();
+            }
+          } else {
+            logger.info('Bootstrapping key from config');
+
+            credentials.key = signingOptions?.key;
+          }
+
+          if (fs.existsSync(signingOptions?.cert)) {
+            if (fs.lstatSync(signingOptions?.cert).isFile()) {
+              logger.info('Bootstrapping cert from file');
+
+              credentials.cert = fs
+                .readFileSync(signingOptions?.cert)
+                .toString();
+            }
+          } else {
+            logger.info('Bootstrapping cert from config');
+
+            credentials.cert = signingOptions?.cert;
+          }
+        }
+
+        if (credentials.key === '' || credentials.cert === '') {
+          throw new Error('Invalid credentials');
+        }
+
+        server = https.createServer(credentials, app) as http.Server;
+      } else {
```

Comment:
this is a pretty large `if`, maybe break out into `createHttpsServer` and `createHttpServer` functions?

Reply:
Created a hostFactory module which exports two functions. :)